File: studio-sdk/python/studio_sdk/services/studioApiService.py

import aiohttp
import json


def get_service_by_quality_and_lang(services, quality, lang):
    return next((s for s in services if s.get("language") in ("*", lang)), None)


def generate_service_config(
    service,
    enable_punctuation=False,
    enable_diarization=False,
    number_of_speaker=0,
    language_value=None,
):
    if not service:
        return None

    language_value = language_value or service.get("language", "*")
    is_whisper = service.get("model_type") == "whisper"
    sub_services = service.get("sub_services", {})

    punctuation_list = sub_services.get("punctuation", [])
    punctuation_service = (
        punctuation_list[0]["service_name"]
        if enable_punctuation and not is_whisper and punctuation_list
        else None
    )

    diarization_list = sub_services.get("diarization", [])
    diarization_service = (
        diarization_list[0]["service_name"]
        if enable_diarization and diarization_list
        else None
    )

    return {
        "serviceName": service["serviceName"],
        "endpoint": remove_leading_slash(service["endpoints"][0]["endpoint"]),
        "lang": language_value,
        "config": {
            "language": language_value,
            "punctuationConfig": {
                "enablePunctuation": enable_punctuation and not is_whisper,
                "serviceName": punctuation_service,
            },
            "diarizationConfig": {
                "enableDiarization": enable_diarization,
                "numberOfSpeaker": (
                    int(number_of_speaker)
                    if enable_diarization and number_of_speaker > 0
                    else None
                ),
                "maxNumberOfSpeaker": 100 if enable_diarization else None,
                "serviceName": diarization_service,
            },
            "enableNormalization": True,
            "modelType": service.get("model_type"),
            "vadConfig": {
                "enableVAD": True,
                "methodName": "WebRTC",
                "minDuration": 30 if is_whisper else 0,
            },
        },
    }


def remove_leading_slash(s: str) -> str:
    return s.lstrip("/")


# --- Decorators ---

def with_token(method):
    async def wrapper(self, *args, **kwargs):
        if "token" not in kwargs:
            if not self.token:
                raise RuntimeError(
                    "No token. You need to pass a token or login first")
            kwargs["token"] = self.token
        return await method(self, *args, **kwargs)
    return wrapper


def with_organization_id(method):
    async def wrapper(self, *args, **kwargs):
        if "organizationId" not in kwargs:
            if not self.organizations:
                await self.fetch_organizations()
            if not self.organizations:
                raise RuntimeError("User has no organizations")
            kwargs["organizationId"] = self.organizations[0]["_id"]
        return await method(self, *args, **kwargs)
    return wrapper


def with_upload_config(method):
    async def wrapper(self, *args, **kwargs):
        file = kwargs.pop("file", None)
        lang = kwargs.get("lang", "*")

        if not self.asr_services:
            await self.fetch_asr_services()

        selected = get_service_by_quality_and_lang(self.asr_services, 1, lang)
        if not selected:
            raise RuntimeError(f"No ASR services available for lang={lang}")

        service_config = generate_service_config(
            selected,
            enable_punctuation=kwargs.get("enablePunctuation", False),
            enable_diarization=kwargs.get("enableDiarization", False),
            number_of_speaker=kwargs.get("numberOfSpeaker", 0),
        )

        kwargs.setdefault("serviceName", service_config["serviceName"])
        kwargs.setdefault("endpoint", service_config["endpoint"])
        kwargs.setdefault("transcriptionConfig", service_config["config"])
        kwargs.setdefault("lang", lang)
        kwargs["file"] = file
        kwargs.setdefault("segmentCharSize", 2000)
        return await method(self, *args, **kwargs)
    return wrapper


class StudioApiService:
    def __init__(
        self,
        base_url="https://studio.linto.ai",
        api_path="cm-api",
        auth_path="auth",
        token=None,
    ):
        self.base_api_url = f"{base_url}/{api_path}"
        self.base_auth_url = f"{base_url}/{auth_path}"
        self.token = token
        self.organizations = []
        self.asr_services = []

    # --- API methods ---

    @with_token
    async def fetch_asr_services(self, **kwargs):
        services = await self._fetch_services(**kwargs)
        self.asr_services = services
        return [s for s in services if "asr" in s.get("scope", [])]

    @with_token
    async def get_media_status(self, mediaId: str, **kwargs):
        return await self._fetch_media(mediaId=mediaId, key="job", **kwargs)

    @with_token
    async def get_media(self, mediaId: str, **kwargs):
        return await self._fetch_media(mediaId=mediaId, **kwargs)

    @with_token
    async def fetch_organizations(self, **kwargs):
        organizations = await self._fetch_organizations(**kwargs)
        self.organizations = organizations
        return organizations

    @with_token
    @with_organization_id
    @with_upload_config
    async def upload_file(
        self,
        file,
        name: str = "name",
        lang: str = "*",
        enablePunctuation: bool = False,
        enableDiarization: bool = False,
        numberOfSpeaker: int = 0,
        segmentCharSize: int = 2000,
        **kwargs,
    ):
        """
        Upload a file for transcription.
        """
        return await self._upload_file(
            file=file,
            name=name,
            lang=lang,
            enablePunctuation=enablePunctuation,
            enableDiarization=enableDiarization,
            numberOfSpeaker=numberOfSpeaker,
            segmentCharSize=segmentCharSize,
            **kwargs,
        )

    async def login(self, email: str, password: str):
        async with aiohttp.ClientSession() as session:
            async with session.post(
                f"{self.base_auth_url}/login",
                json={"email": email, "password": password},
            ) as resp:
                res = await resp.json()
                self.token = res.get("auth_token")
                return res

    # --- Internal fetchers ---

    async def _fetch_media(self, mediaId, **kwargs):
        url = f"{self.base_api_url}/conversations/{mediaId}"
        return await self._send_request("GET", url, **kwargs)

    async def _fetch_services(self, **kwargs):
        url = f"{self.base_api_url}/services"
        return await self._send_request("GET", url, **kwargs)

    async def _fetch_organizations(self, **kwargs):
        url = f"{self.base_api_url}/organizations"
        return await self._send_request("GET", url, **kwargs)

    async def _upload_file(self, **kwargs):
        file = kwargs.get("file")
        if not file:
            raise RuntimeError("File is required")

        org_id = kwargs["organizationId"]
        url = f"{self.base_api_url}/organizations/{org_id}/conversations/create"

        form = aiohttp.FormData()
        form.add_field("name", kwargs.get("name", "name"))
        form.add_field("file", file)
        form.add_field("serviceName", kwargs["serviceName"])

        print("conf", str(
            kwargs["transcriptionConfig"]))

        form.add_field("transcriptionConfig", json.dumps(
            kwargs["transcriptionConfig"]))
        form.add_field("segmentCharSize", json.dumps(
            kwargs["segmentCharSize"]))
        form.add_field("lang", kwargs["lang"])
        form.add_field("endpoint", kwargs["endpoint"])

        return await self._send_request("POST", url, data=form, token=kwargs["token"])

    # --- Generic request sender ---

    async def _send_request(self, method, url, **kwargs):
        headers = {"Authorization": f"Bearer {kwargs.get('token')}"}
        async with aiohttp.ClientSession() as session:
            print(url)
            async with session.request(
                method, url, headers=headers, json=kwargs.get("json"), data=kwargs.get("data")
            ) as resp:
                return await resp.json()

File: studio-sdk/javascript/src/services/studioApiService.js
import {
  prepareRequest,
  prepareMultipartFormData,
  sendRequest,
} from "../request.js"

import mediaFactory from "../models/media.js"

export class StudioApiService {
  constructor({ baseUrl = "https://studio.linto.ai", token = null }) {
    this.baseApiUrl = baseUrl + "/api"
    this.baseAuthUrl = baseUrl + "/auth"
    this.token = token
    this.organizations = []
    this.asrServices = []
  }

  async fetchAsrServices(args) {
    const services = await this.#withToken(this.#fetchServices)(args)
    this.asrServices = services
    return services?.filter((service) => service.scope.indexOf("stt") > -1)
  }

  async getMediaStatus({ mediaId }) {
    const conv = await this.#withToken(this.#fetchMedia)({
      mediaId,
      key: "job",
    })
    return conv
  }

  async getMedia({ mediaId }) {
    const conv = await this.#withToken(this.#fetchMedia)({
      mediaId,
    })
    return mediaFactory(conv)
  }

  async fetchOrganizations(args) {
    const organizations = await this.#withToken(this.#fetchOrganizations)(args)
    this.organizations = organizations
    return organizations
  }

  async uploadFile(args) {
    return await this.#withToken(
      this.#withOrganizationId(this.#withUploadConfig(this.#uploadFile))
    )(args)
  }

  // -- LLM / Summary methods --

  async fetchLlmServices(args) {
    return await this.#withToken(
      this.#withOrganizationId(this.#fetchLlmServices.bind(this))
    )(args)
  }

  async triggerSummary({ conversationId, format, flavor } = {}) {
    return await this.#withToken(this.#triggerSummary)({
      conversationId,
      format,
      flavor,
    })
  }

  async getExportList({ conversationId } = {}) {
    return await this.#withToken(this.#getExportList)({ conversationId })
  }

  async getExportContent({ conversationId, jobId } = {}) {
    return await this.#withToken(this.#getExportContent)({
      conversationId,
      jobId,
    })
  }

  // -- Download / Publication methods --

  async downloadConversation({ conversationId, format = "docx" } = {}) {
    return await this.#withToken(this.#downloadConversation)({
      conversationId,
      format,
    })
  }

  async getPublicationTemplates(args = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#getPublicationTemplates.bind(this))
    )(args)
  }

  async getTemplatePlaceholders({ templateId } = {}) {
    return await this.#withToken(this.#getTemplatePlaceholders)({
      templateId,
    })
  }

  async exportWithTemplate({
    jobId,
    format = "pdf",
    templateId,
    versionNumber,
  } = {}) {
    return await this.#withToken(this.#exportWithTemplate)({
      jobId,
      format,
      templateId,
      versionNumber,
    })
  }

  // -- Taxonomy methods (categories, tags, folders) --

  async listCategories(args = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#listCategories.bind(this))
    )(args)
  }

  async listTags({ categoryId } = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#listTags.bind(this))
    )({ categoryId })
  }

  async createTag({ categoryId, name, color, emoji, description } = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#createTag.bind(this))
    )({ categoryId, name, color, emoji, description })
  }

  async listFolders({ tree = false, withConversationCount = false } = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#listFolders.bind(this))
    )({ tree, withConversationCount })
  }

  async createFolder({
    name,
    parentId,
    color,
    emoji,
    visibility = "public",
  } = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#createFolder.bind(this))
    )({ name, parentId, color, emoji, visibility })
  }

  async moveConversationToFolder({ folderId, conversationId } = {}) {
    return await this.#withToken(
      this.#withOrganizationId(this.#moveConversationToFolder.bind(this))
    )({ folderId, conversationId })
  }

  async getConversation({ conversationId } = {}) {
    return await this.#withToken(this.#getConversation)({ conversationId })
  }

  async searchUsers({ search } = {}) {
    return await this.#withToken(this.#searchUsers)({ search })
  }

  async updateConversation({ conversationId, data } = {}) {
    return await this.#withToken(this.#updateConversation)({
      conversationId,
      data,
    })
  }

  async shareConversation({
    conversationId,
    email,
    right = 1,
    notify = true,
  } = {}) {
    return await this.#withToken(this.#shareConversation)({
      conversationId,
      email,
      right,
      notify,
    })
  }

  async login({ email, password }) {
    const req = prepareRequest(`${this.baseAuthUrl}/login`, "POST", {
      email,
      password,
    })

    const res = await sendRequest(req)
    this.token = res.auth_token
    return res
  }

  // -- Decorators --

  #withOrganizationId(method) {
    return async (args = {}) => {
      if (!args?.organizationId) {
        if (this.organizations.length === 0) {
          await this.fetchOrganizations()
        }

        if (this.organizations.length === 0) {
          throw new Error("User has no organizations")
        }

        args.organizationId = this.organizations[0]._id
      }

      return await method(args)
    }
  }

  #withToken(method) {
    return async (args = {}) => {
      if (!args?.token) {
        if (!this.token) {
          throw new Error("No token. You need to pass a token or login first")
        }

        args.token = this.token
      }
      return await method.bind(this)(args)
    }
  }

  #withUploadConfig(method) {
    return async ({ file, lang = "*", ...args } = {}) => {
      if (this.asrServices.length == 0) {
        await this.fetchAsrServices()
      }

      const selectedService = getServiceByQualityAndLang(
        this.asrServices,
        1,
        lang
      )

      if (!selectedService) {
        throw new Error(`No ASR services available for lang=${lang}`)
      }

      const serviceConfig = generateServiceConfig(selectedService, {
        enablePunctuation: args["enablePunctuation"],
        enableDiarization: args["enableDiarization"],
        numberOfSpeaker: args["numberOfSpeaker"],
      })

      args["serviceName"] = args["serviceName"] ?? serviceConfig.serviceName
      args["endpoint"] = args["endpoint"] ?? serviceConfig.endpoint
      args["transcriptionConfig"] =
        args["transcriptionConfig"] ?? serviceConfig.config
      args["lang"] = lang
      args["file"] = file
      args["name"] = args["name"] ?? `imported file ${new Date().toISOString()}`
      args["segmentCharSize"] = args["segmentCharSize"] ?? 2000
      return await method.bind(this)(args)
    }
  }

  // -- API calls --
  async #fetchMedia({ token, mediaId, ...params }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/conversations/${mediaId}`,
      "GET",
      { token, ...params }
    )

    return await sendRequest(req)
  }

  async #fetchServices({ token }) {
    const req = prepareRequest(`${this.baseApiUrl}/services`, "GET", {
      token,
    })

    return await sendRequest(req)
  }

  async #fetchOrganizations({ token }) {
    const req = prepareRequest(`${this.baseApiUrl}/organizations`, "GET", {
      token,
    })

    return await sendRequest(req)
  }

  async #uploadFile({
    token,
    organizationId,
    file,
    name,
    transcriptionConfig,
    serviceName,
    endpoint,
    lang,
    segmentCharSize,
  }) {
    if (!file) {
      throw new Error("File is required")
    }

    let formData = new FormData()
    formData.append("name", name)
    formData.append("file", file)
    formData.append("serviceName", serviceName)
    formData.append("transcriptionConfig", JSON.stringify(transcriptionConfig))
    formData.append("segmentCharSize", segmentCharSize)
    formData.append("lang", lang)
    formData.append("endpoint", endpoint)
    const req = prepareMultipartFormData(
      `${this.baseApiUrl}/organizations/${organizationId}/conversations/create`,
      token,
      formData
    )

    return await sendRequest(req)
  }

  // -- LLM / Summary private implementations --

  async #fetchLlmServices({ token, organizationId }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/services/${organizationId}/llm`,
      "GET",
      { token }
    )

    return await sendRequest(req)
  }

  async #triggerSummary({ token, conversationId, format, flavor }) {
    let url = `${this.baseApiUrl}/conversations/${conversationId}/download?format=${encodeURIComponent(format)}`
    if (flavor) {
      url += `&flavor=${encodeURIComponent(flavor)}`
    }
    const req = prepareRequest(url, "POST", { token })
    return await sendRequest(req)
  }

  async #getExportList({ token, conversationId }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/conversations/${conversationId}/export/list`,
      "GET",
      { token }
    )
    return await sendRequest(req)
  }

  async #getExportContent({ token, conversationId, jobId }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/conversations/${conversationId}/export/${jobId}/content`,
      "GET",
      { token }
    )
    return await sendRequest(req)
  }

  // -- Download / Publication private implementations --

  async #downloadConversation({ token, conversationId, format }) {
    const url = `${this.baseApiUrl}/conversations/${conversationId}/download?format=${encodeURIComponent(format)}`
    const req = prepareRequest(url, "POST", { token })
    const isBinary =
      format === "docx" || format === "odt" || format === "verbatim"
    return await sendRequest(req, {
      responseType: isBinary ? "binary" : undefined,
    })
  }

  async #getPublicationTemplates({ token, organizationId }) {
    const url = `${this.baseApiUrl}/publication/templates?organization_id=${encodeURIComponent(organizationId)}`
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req)
  }

  async #getTemplatePlaceholders({ token, templateId }) {
    const url = `${this.baseApiUrl}/publication/templates/${templateId}/placeholders`
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req)
  }

  async #exportWithTemplate({
    token,
    jobId,
    format,
    templateId,
    versionNumber,
  }) {
    let url = `${this.baseApiUrl}/publication/${jobId}/export/${format}`
    const params = []
    if (templateId !== undefined && templateId !== null && templateId !== "") {
      params.push(`templateId=${encodeURIComponent(templateId)}`)
    }
    if (versionNumber !== undefined && versionNumber !== null) {
      params.push(`versionNumber=${encodeURIComponent(versionNumber)}`)
    }
    if (params.length > 0) {
      url += `?${params.join("&")}`
    }
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req, { responseType: "binary" })
  }

  // -- Taxonomy private implementations --

  async #listCategories({ token, organizationId }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/organizations/${organizationId}/categories`,
      "GET",
      { token }
    )
    return await sendRequest(req)
  }

  async #listTags({ token, organizationId, categoryId }) {
    const url = `${this.baseApiUrl}/organizations/${organizationId}/tags?categoryId=${encodeURIComponent(categoryId)}`
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req)
  }

  async #createTag({
    token,
    organizationId,
    categoryId,
    name,
    color,
    emoji,
    description,
  }) {
    const payload = {
      organizationId,
      categoryId,
      name,
    }
    if (color !== undefined && color !== null) payload.color = color
    if (emoji !== undefined && emoji !== null) payload.emoji = emoji
    if (description !== undefined && description !== null)
      payload.description = description

    const req = prepareRequest(
      `${this.baseApiUrl}/organizations/${organizationId}/tags`,
      "POST",
      { token, ...payload }
    )
    return await sendRequest(req)
  }

  async #listFolders({ token, organizationId, tree, withConversationCount }) {
    const params = []
    if (tree) params.push("tree=true")
    if (withConversationCount) params.push("withConversationCount=true")
    let url = `${this.baseApiUrl}/organizations/${organizationId}/folders`
    if (params.length > 0) {
      url += "?" + params.join("&")
    }
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req)
  }

  async #createFolder({
    token,
    organizationId,
    name,
    parentId,
    color,
    emoji,
    visibility,
  }) {
    const payload = { name, visibility }
    if (parentId !== undefined && parentId !== null) payload.parentId = parentId
    if (color !== undefined && color !== null) payload.color = color
    if (emoji !== undefined && emoji !== null) payload.emoji = emoji

    const req = prepareRequest(
      `${this.baseApiUrl}/organizations/${organizationId}/folders`,
      "POST",
      { token, ...payload }
    )
    return await sendRequest(req)
  }

  async #moveConversationToFolder({
    token,
    organizationId,
    folderId,
    conversationId,
  }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/organizations/${organizationId}/folders/${folderId}/conversations/${conversationId}`,
      "POST",
      { token }
    )
    return await sendRequest(req)
  }

  async #getConversation({ token, conversationId }) {
    const req = prepareRequest(
      `${this.baseApiUrl}/conversations/${conversationId}`,
      "GET",
      { token }
    )
    return await sendRequest(req)
  }

  async #searchUsers({ token, search }) {
    const url = `${this.baseApiUrl}/users/search?search=${encodeURIComponent(search)}`
    const req = prepareRequest(url, "GET", { token })
    return await sendRequest(req)
  }

  async #updateConversation({ token, conversationId, data }) {
    const url = `${this.baseApiUrl}/conversations/${conversationId}`
    const requestObj = new Request(url, {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
        Authorization: token ? `Bearer ${token}` : null,
      },
      body: JSON.stringify(data ?? {}),
    })
    return await sendRequest(requestObj)
  }

  async #shareConversation({ token, conversationId, email, right, notify }) {
    const url = `${this.baseApiUrl}/conversations/${conversationId}/invite`
    const payload = { email, right }
    if (notify === false) {
      payload.notify = false
    }
    const requestObj = new Request(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: token ? `Bearer ${token}` : null,
      },
      body: JSON.stringify(payload),
    })
    return await sendRequest(requestObj)
  }
}

function getServiceByQualityAndLang(services, quality, lang) {
  return services.find((service) => {
    return service.language == "*" || service.language == lang
  })
}

function generateServiceConfig(
  service,
  {
    enablePunctuation = false,
    enableDiarization = false,
    numberOfSpeaker = 0,
    languageValue = service?.language || "*",
  } = {}
) {
  const isWhisper = service?.model_type === "whisper"
  const subServices = service?.sub_services

  const punctuationServiceList = subServices?.punctuation
  const punctuationService =
    enablePunctuation && !isWhisper && punctuationServiceList.length > 0
      ? punctuationServiceList[0].service_name
      : null

  const diarizationServiceList = subServices?.diarization
  const diarizationService =
    enableDiarization && diarizationServiceList.length > 0
      ? diarizationServiceList[0].service_name
      : null

  return {
    serviceName: service.serviceName,
    endpoint: removeLeadingSlash(service.endpoints[0].endpoint),
    lang: languageValue,
    config: {
      language: languageValue,
      punctuationConfig: {
        enablePunctuation: enablePunctuation && !isWhisper,
        serviceName: punctuationService,
      },
      diarizationConfig: {
        enableDiarization: enableDiarization,
        numberOfSpeaker:
          enableDiarization && numberOfSpeaker > 0
            ? parseInt(numberOfSpeaker)
            : null,
        maxNumberOfSpeaker: enableDiarization ? 100 : null,
        serviceName: diarizationService,
      },
      enableNormalization: true,
      modelType: service.model_type,
      vadConfig: isWhisper
        ? {
            enableVAD: true,
            methodName: "WebRTC",
            minDuration: 30,
          }
        : {
            enableVAD: true,
            methodName: "WebRTC",
            minDuration: 0,
          },
    },
  }
}

function removeLeadingSlash(str) {
  return str.replace(/^\/+/, "")
}
